File: 2024/haskell/day06b/Main.hs
module Main ( main ) where

import Data.Array ( Array )
import qualified Data.Array as Array
import Data.Bifunctor ( bimap )
import Data.Set ( Set )
import qualified Data.Set as Set
import System.Environment ( getArgs, getProgName )
import System.Exit ( exitFailure )

type Position = (Int, Int)

data Lab = Lab { start :: Position
               , direction :: Int
               , grid :: Array (Int, Int) Char
               , numRows :: Int
               , numCols :: Int
               }
               deriving (Eq, Show)

usage :: IO ()
usage = do
  progname <- getProgName
  putStrLn $ "usage: " ++ progname ++ " <file>"
  exitFailure

parse :: String -> Lab
parse contents =
    let rows = lines contents
        numRows = length rows
        numCols = length (head rows)
        start = head $ [(r, c) | r <- [0..(numRows - 1)], c <- [0..(numCols - 1)], rows !! r !! c == '^']
        elems = [((i, j), c) | i <- [0..numRows - 1], j <- [0..numCols - 1], let c = (rows !! i) !! j]
        grid = Array.array ((0, 0), (numRows - 1, numCols - 1)) elems
    in Lab { start = start, direction = 0, grid = grid, numRows = numRows, numCols = numCols }

directions :: [(Int, Int)]
directions = [(-1, 0), (0, 1), (1, 0), (0, -1)]

inBounds :: Lab -> Position -> Bool
inBounds lab pos@(r, c) = r >= 0 && r < numRows lab && c >= 0 && c < numCols lab

guardPositions :: Lab -> [(Position, Int)]
guardPositions lab =
    let pos = start lab
        dir = direction lab
        guardPositions' pos dir positions =
            let positions' = positions ++ [(pos, dir)]
                nextPos = bimap (fst pos +) (snd pos +) (directions !! dir)
            in if not (inBounds lab nextPos)
                then positions'
                else let dir' = (if grid lab Array.! nextPos == '#'
                                    then (dir + 1) `mod` 4
                                    else dir)
                         pos' = bimap (fst pos +) (snd pos +) (directions !! dir')
                     in guardPositions' pos' dir' positions'
    in guardPositions' pos dir []

hasLoop :: Lab -> Bool
hasLoop !lab = hasLoop' Set.empty (start lab) (direction lab) False
    where
        hasLoop' positions pos dir turn =
            Set.member (pos, dir) positions || (let positions' = Set.insert (pos, dir) positions
                                                    nextPos = bimap (fst pos +) (snd pos +) (directions !! dir)
                                                in (inBounds lab nextPos && (if grid lab Array.! nextPos == '#'
                                                       then hasLoop' positions' pos ((dir + 1) `mod` 4) True
                                                       else hasLoop' positions' nextPos dir False)))

numObstructions :: Lab -> Int
numObstructions lab =
    let route = Set.fromList $ map fst $ guardPositions lab
        loopPos = foldl
                    (\loopPos' pos ->
                        let grid' = grid lab Array.// [(pos, '#')]
                            lab' = Lab { start = start lab
                                       , direction = direction lab
                                       , grid = grid'
                                       , numRows = numRows lab
                                       , numCols = numCols lab
                                       }
                        in if hasLoop lab'
                            then Set.insert pos loopPos'
                            else loopPos'
                    )
                    Set.empty
                    route
    in Set.size loopPos

process :: String -> Int
process contents = numObstructions (parse contents)

main :: IO ()
main = do
  args <- getArgs
  case args of
    [filename] -> do
      contents <- readFile filename
      let result = process contents
      putStrLn ("result = " ++ show result)
    _ -> usage

File: 2024/rust/day06b/src/main.rs
use std::collections::HashSet;
use std::env;
use std::fs;
use std::process;

fn usage() {
    let progname = env::args().next().unwrap();
    println!("usage: {progname} <file>");
    process::exit(1);
}

fn build_grid(contents: &str) -> Vec<Vec<char>> {
    let mut grid: Vec<Vec<char>> = vec![];
    for line in contents.lines() {
        let row: Vec<char> = line.chars().collect();
        grid.push(row);
    }
    grid
}

fn find_start(grid: &[Vec<char>]) -> Option<(usize, usize)> {
    let extents = (grid.len(), grid[0].len());
    for (row, row_chars) in grid.iter().enumerate().take(extents.0) {
        for (col, ch) in row_chars.iter().enumerate().take(extents.1) {
            if *ch == '^' {
                return Some((row, col));
            }
        }
    }
    None
}

fn find_obstacles(grid: &[Vec<char>]) -> HashSet<(usize, usize)> {
    let mut obstacles: HashSet<(usize, usize)> = HashSet::new();
    let extents = (grid.len(), grid[0].len());
    for (row, row_chars) in grid.iter().enumerate().take(extents.0) {
        for (col, ch) in row_chars.iter().enumerate().take(extents.1) {
            if *ch == '#' {
                obstacles.insert((row, col));
            }
        }
    }
    obstacles
}

fn turn(dir: &(isize, isize)) -> (isize, isize) {
    match dir {
        (-1, 0) => (0, 1),
        (0, 1) => (1, 0),
        (1, 0) => (0, -1),
        (0, -1) => (-1, 0),
        _ => panic!("unknown direction in turn()"),
    }
}

fn in_bounds(pos: &(isize, isize), extents: &(usize, usize)) -> bool {
    pos.0 >= 0 && pos.0 < extents.0 as isize && pos.1 >= 0 && pos.1 < extents.1 as isize
}

fn walk(
    start_pos: &(usize, usize),
    start_dir: &(isize, isize),
    extents: &(usize, usize),
    obstacles: &HashSet<(usize, usize)>,
    check_cycles: bool,
) -> Vec<((usize, usize), (isize, isize))> {
    let mut path: Vec<((usize, usize), (isize, isize))> = vec![];
    let mut pos = *start_pos;
    let mut dir = *start_dir;
    loop {
        path.push((pos, dir));
        let maybe_pos = (pos.0 as isize + dir.0, pos.1 as isize + dir.1);
        if !in_bounds(&maybe_pos, extents) {
            break;
        } else if check_cycles && path.iter().filter(|&&x| x == *path.last().unwrap()).count() > 1 {
            break;
        }
        let new_pos = (maybe_pos.0 as usize, maybe_pos.1 as usize);
        if obstacles.contains(&new_pos) {
            dir = turn(&dir);
        } else {
            pos = new_pos;
        }
    }
    path
}

fn try_obstacle(
    pos: &(usize, usize),
    dir: &(isize, isize),
    extents: &(usize, usize),
    obstacles: &HashSet<(usize, usize)>,
    new_obstacle: &(usize, usize),
) -> bool {
    let mut new_obstacles = obstacles.clone();
    new_obstacles.insert(*new_obstacle);
    let sim_path = walk(pos, dir, extents, &new_obstacles, true);
    let last_state = sim_path.last().unwrap();
    let count = sim_path.iter().filter(|&&x| x == *last_state).count();
    count > 1
}

fn add_obstacles(
    path: &[((usize, usize), (isize, isize))],
    start_pos: &(usize, usize),
    start_dir: &(isize, isize),
    extents: &(usize, usize),
    obstacles: &HashSet<(usize, usize)>,
) -> HashSet<(usize, usize)> {
    let mut added_obstacles: HashSet<(usize, usize)> = HashSet::new();
    for &current_state in path.iter().skip(1) {
        let new_obstacle = current_state.0;
        if try_obstacle(&start_pos, &start_dir, extents, obstacles, &new_obstacle) {
            added_obstacles.insert(new_obstacle);
        }
    }
    added_obstacles
}

fn process(contents: &str) -> u32 {
    let grid = build_grid(contents);
    let extents = (grid.len(), grid[0].len());
    let start_pos = match find_start(&grid) {
        Some(start) => start,
        None => panic!("start position not found in data!"),
    };
    let start_dir = (-1, 0);
    let obstacles = find_obstacles(&grid);
    let path = walk(&start_pos, &start_dir, &extents, &obstacles, false);
    add_obstacles(&path, &start_pos, &start_dir, &extents, &obstacles).len() as u32
}

fn main() {
    if env::args().count() < 2 {
        usage();
    }
    let filename = env::args().nth(1).unwrap();
    let contents = fs::read_to_string(filename).expect("read of input file failed");
    let result = process(&contents);
    println!("result = {result}");
}
